[Static Analyzer CI] Filter unexpected results using results.webkit.org to reduce build times on EWS

briannafan opened the pull request:
#### 94fb4893473fce89ffca8bfc7a9dcab2842a14f6
<pre>
[Static Analyzer CI] Filter unexpected results using results.webkit.org to reduce build times on EWS
<a href="https://bugs.webkit.org/show_bug.cgi?id=282770">https://bugs.webkit.org/show_bug.cgi?id=282770</a>
<a href="https://rdar.apple.com/139393719">rdar://139393719</a>

Reviewed by Jonathan Bedard.

Consult results.webkit.org on the first run of FindUnexpectedStaticAnalyzerResults to determine if
failures and successes on the PR are pre-existing. If so, we can filter those results out.

* Tools/CISupport/ews-build/steps.py:

(FindUnexpectedStaticAnalyzerResults):
    - If self.expectations = True (1st run on PR):
          - Check results database for each unexpected result
           1. If it fails, fall back to rebuilding on ToT.
           2. If it succeeds with no changes, run GenerateSaferCPPResultsIndex, Delete, Archive,
              Upload, Extract, &amp; Display results.
           3. If it succeeds with changes, run DownloadUnexpectedResultsFromMaster, DeleteStaticAnalyzerResults,
              GenerateSaferCPPResultsIndex, then Archive, Upload, Extract, &amp; Display results.
    - If self.expectations = False (2nd run on ToT):
          - These are not new changes and are already running.
          - Compare the PR and ToT results. Then run Archive, Upload, Extract, &amp; Display results.
          - Upload ToT results to results.webkit.org.

(FindUnexpectedStaticAnalyzerResults.run): Add new logic for each case enumerated above.
(FindUnexpectedStaticAnalyzerResults.filter_results_using_results_db): Added.
(FindUnexpectedStaticAnalyzerResults.check_results_db): Added.
(FindUnexpectedStaticAnalyzerResults.find_unexpected_results): Added.
(FindUnexpectedStaticAnalyzerResults.createResultMessage): Added.

Canonical link: <a href="https://commits.webkit.org/287360@main">https://commits.webkit.org/287360@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/200ba8ddb746a8f84523926cd127e500e9d1abbd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79137 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58170 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32513 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83769 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30339 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81270 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67271 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6436 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61921 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19838 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82204 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51973 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72116 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42225 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49324 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26236 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28710 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70440 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26658 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85159 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6450 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4463 "Found 1 new test failure: imported/w3c/web-platform-tests/mediacapture-record/MediaRecorder-peerconnection.https.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70166 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/78827 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6613 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67988 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69414 "Found 3 new API test failures: /TestWebKit:WebKit.LoadCanceledNoServerRedirectCallback, /TestWebKit:WebKit.GeolocationBasicWithHighAccuracy, /TestWebKit:WebKit2UserMessageRoundTripTest.WKURL (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13464 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12264 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12246 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6403 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6352 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9812 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8144 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

briannafan commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -6040,7 +6040,7 @@ def getResultSummary(self):
 
 
-class PrintClangVersion(shell.ShellCommandNewStyle):
+class PrintClangVersion(shell.ShellCommandNewStyle, AddToLogMixin):
```
Left over from debugging, I'll delete in the next commit.

ryanhaddad commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7302,43 +7310,170 @@ def run(self):
-    def createResultMessage(self):
+    @defer.inlineCallbacks
+    def write_unexpected_results_file(self):
+        resultDirectory = f"public_html/results/{self.getProperty('buildername')}"
```
Should we be concerned about this path being written / accessed by multiple `macOS-Safer-CPP-Checks-EWS` builds simultaneously?

briannafan replied:
Ah. Great point - let me modify this function so it creates a directory for each build.

briannafan replied:
Here's a successful run after changing the directory: https://ews-build.webkit-dev.org/#/builders/218/builds/1741

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7040,31 +7040,39 @@ class ParseStaticAnalyzerResultsWithoutChange(ParseStaticAnalyzerResults):
+    suite = 'safer-cpp-checks'
 
-    def __init__(self, expectations=False, **kwargs):
+    def __init__(self, expectations=False, filtered_changes=False, **kwargs):
```
What does "filtered_changes" mean? I'm a little surprised to see that as a boolean, instead of a list.

briannafan replied:
I guess it could be named `was_filtered` or something to that effect? Essentially I want to track whether or not going through results db actually changed the results.

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7084,43 +7092,169 @@ def run(self):
 
-        self.createResultMessage()
+        self.find_unexpected_results()
+        unexpected_results = self.getProperty('num_failing_files', 0) or self.getProperty('num_unexpected_issues', 0) or self.getProperty('num_passing_files', 0)
```
Should probably be `num_unexpected_results`, since it's going to be a number, not a list.

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7040,31 +7040,39 @@ class ParseStaticAnalyzerResultsWithoutChange(ParseStaticAnalyzerResults):
 
-    def __init__(self, expectations=False, **kwargs):
+    def __init__(self, expectations=False, filtered_changes=False, **kwargs):
         self.expectations = expectations  # If true, results will be compared against checked-in expectations. Otherwise, they're compared against a previous run.
```
This should probably be named something like `have_expectation` to make it clear it's boolean.

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7084,43 +7092,169 @@ def run(self):
+        self.find_unexpected_results()
+        unexpected_results = self.getProperty('num_failing_files', 0) or self.getProperty('num_unexpected_issues', 0) or self.getProperty('num_passing_files', 0)
+        unexpected_results_after_filter = None
+        if self.expectations and unexpected_results:  # Only consult results database on the first run
```
Isn't `self.expectations = False` by default? I would expect the first run to be the "default" run.

briannafan replied:
That's true it would be less confusing. Made the change.

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7084,43 +7092,169 @@ def run(self):
+            logTextJson = self.log_observer_json.getStdout()
+            yield self._addToLog('stdio', f'Checking results database for unexpected results...\n')
+            successful_filter = yield self.filter_results_using_results_db(logTextJson)
+            unexpected_results_after_filter = self.getProperty('num_failing_files', 0) or self.getProperty('num_unexpected_issues', 0) or self.getProperty('num_passing_files', 0)
```
Does `self.filter_results_using_results_db` set properties? That's a bit surprising since it's `self.`, usually we reserve properties to propagate data between steps, not within the same step.

briannafan replied:
It does set properties. The reason I did this was because `DisplaySaferCPPResults` uses these properties, so I thought they could be used intra-step since we're setting them anyways

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7084,43 +7092,169 @@ def run(self):
+        if self.expectations and unexpected_results:  # Only consult results database on the first run
+            logTextJson = self.log_observer_json.getStdout()
+            yield self._addToLog('stdio', f'Checking results database for unexpected results...\n')
+            successful_filter = yield self.filter_results_using_results_db(logTextJson)
```
What does `filter_results_using_results_db`? I don't think there is a "filter" object here. Is it a boolean?

briannafan replied:
Yeah it returns a boolean if checking results db is successful. It's similar to `filter_failures_using_results_db` but does it for failures and passes instead of only failures.

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7084,43 +7092,169 @@ def run(self):
+        if not self.filtered_changes:
+            self.result_message += f"{new_issues} new issue{'s' if new_issues > 1 else ''} " if new_issues else ''
+        self.result_message += f"{failing_files} failing file{'s' if failing_files > 1 else ''} " if failing_files else ''
+        self.result_message += f"{fixed_files} fixed file{'s' if fixed_files > 1 else ''}" if fixed_files else ''
```
Is appending the right idiom here? Seems like we should be setting, or perhaps returning and having the caller own setting.

JonWBedard commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -7084,43 +7092,169 @@ def run(self):
         return defer.returnValue(rc)
 
-    def createResultMessage(self):
+    def write_unexpected_results_file(self):
```
This writes to the master. Might want to make that explicit in the function name, some might find that surprising.

briannafan committed:
[Static Analyzer CI] Filter unexpected results using results.webkit.org to reduce build times on EWS
https://bugs.webkit.org/show_bug.cgi?id=282770
rdar://139393719

Reviewed by Jonathan Bedard.

Consult results.webkit.org on the first run of FindUnexpectedStaticAnalyzerResults to determine if
failures and successes on the PR are pre-existing. If so, we can filter those results out.

* Tools/CISupport/ews-build/steps.py:

(FindUnexpectedStaticAnalyzerResults):
    - If self.expectations = True (1st run on PR):
          - Check results database for each unexpected result
           1. If it fails, fall back to rebuilding on ToT.
           2. If it succeeds with no changes, run GenerateSaferCPPResultsIndex, Delete, Archive,
              Upload, Extract, & Display results.
           3. If it succeeds with changes, run DownloadUnexpectedResultsFromMaster, DeleteStaticAnalyzerResults,
              GenerateSaferCPPResultsIndex, then Archive, Upload, Extract, & Display results.
    - If self.expectations = False (2nd run on ToT):
          - These are not new changes and are already running.
          - Compare the PR and ToT results. Then run Archive, Upload, Extract, & Display results.
          - Upload ToT results to results.webkit.org.

(FindUnexpectedStaticAnalyzerResults.run): Add new logic for each case enumerated above.
(FindUnexpectedStaticAnalyzerResults.filter_results_using_results_db): Added.
(FindUnexpectedStaticAnalyzerResults.check_results_db): Added.
(FindUnexpectedStaticAnalyzerResults.find_unexpected_results): Added.
(FindUnexpectedStaticAnalyzerResults.createResultMessage): Added.

Canonical link: https://commits.webkit.org/287360@main
